cpp/src/arrow/array-binary-test.cc

```diff
@@ -85,268 +94,132 @@ class TestStringArray : public ::testing::Test {
     }
   }
-  Done();
-
-  ASSERT_EQ(reps * N, result_->length());
-  ASSERT_EQ(reps, result_->null_count());
-  ASSERT_EQ(reps * 6, result_->value_data()->size());
 
-  CheckStringArray(*result_, strings, is_valid, reps);
-}
-
-TEST_F(TestStringBuilder, TestAppendVector) {
-  std::vector<std::string> strings = {"", "bb", "a", "", "ccc"};
-  std::vector<uint8_t> valid_bytes = {1, 1, 1, 0, 1};
-
-  int N = static_cast<int>(strings.size());
-  int reps = 1000;
-
-  for (int j = 0; j < reps; ++j) {
-    ASSERT_OK(builder_->AppendValues(strings, valid_bytes.data()));
-  }
-  Done();
-
-  ASSERT_EQ(reps * N, result_->length());
-  ASSERT_EQ(reps, result_->null_count());
-  ASSERT_EQ(reps * 6, result_->value_data()->size());
-
-  CheckStringArray(*result_, strings, valid_bytes, reps);
-}
-
-TEST_F(TestStringBuilder, TestAppendCStringsWithValidBytes) {
-  const char* strings[] = {nullptr, "aaa", nullptr, "ignored", ""};
-  std::vector<uint8_t> valid_bytes = {1, 1, 1, 0, 1};
-
-  int N = static_cast<int>(sizeof(strings) / sizeof(strings[0]));
-  int reps = 1000;
+  void _TestEmptyStringComparison() {
+    offsets_ = {0, 0, 0, 0, 0, 0};
+    offsets_buf_ = Buffer::Wrap(offsets_);
+    length_ = static_cast<int64_t>(offsets_.size() - 1);
 
-  for (int j = 0; j < reps; ++j) {
-    ASSERT_OK(builder_->AppendValues(strings, N, valid_bytes.data()));
+    auto strings_a = std::make_shared<ArrayType>(length_, offsets_buf_, nullptr,
+                                                 null_bitmap_, null_count_);
+    auto strings_b = std::make_shared<ArrayType>(length_, offsets_buf_, nullptr,
+                                                 null_bitmap_, null_count_);
+    ASSERT_TRUE(strings_a->Equals(strings_b));
   }
-  Done();
 
-  ASSERT_EQ(reps * N, result_->length());
-  ASSERT_EQ(reps * 3, result_->null_count());
-  ASSERT_EQ(reps * 3, result_->value_data()->size());
+  void _TestCompareNullByteSlots() {
+    BuilderType builder;
+    BuilderType builder2;
+    BuilderType builder3;
 
-  CheckStringArray(*result_, {"", "aaa", "", "", ""}, {0, 1, 0, 0, 1}, reps);
-}
+    ASSERT_OK(builder.Append("foo"));
+    ASSERT_OK(builder2.Append("foo"));
+    ASSERT_OK(builder3.Append("foo"));
 
-TEST_F(TestStringBuilder, TestAppendCStringsWithoutValidBytes) {
-  const char* strings[] = {"", "bb", "a", nullptr, "ccc"};
+    ASSERT_OK(builder.Append("bar"));
+    ASSERT_OK(builder2.AppendNull());
 
-  int N = static_cast<int>(sizeof(strings) / sizeof(strings[0]));
-  int reps = 1000;
+    // same length, but different
+    ASSERT_OK(builder3.Append("xyz"));
 
-  for (int j = 0; j < reps; ++j) {
-    ASSERT_OK(builder_->AppendValues(strings, N));
-  }
-  Done();
+    ASSERT_OK(builder.Append("baz"));
+    ASSERT_OK(builder2.Append("baz"));
+    ASSERT_OK(builder3.Append("baz"));
 
-  ASSERT_EQ(reps * N, result_->length());
-  ASSERT_EQ(reps, result_->null_count());
-  ASSERT_EQ(reps * 6, result_->value_data()->size());
+    std::shared_ptr<Array> array, array2, array3;
```

Comment:
You can save yourself the `a<n>` temporaries below by finishing directly to `ArrayType`:
```suggestion
    std::shared_ptr<ArrayType> array, array2, array3;
```

Reply:
The `FinishAndCheckPadding` call wouldn't compile. i'd rather not bother too much here (this is a refactor from previous tests btw).